File: lex.js
var Lexer = function Lexer(string, definitions) {
    var index = 0;
    var length = 1;
    var arrayPosition = -1;
    var _Lexer = this;
    _Lexer.tokens = [];
    _Lexer.lexemes = [];
    _Lexer.token = null;
    _Lexer.lexeme = null;
    _Lexer.bof = true;
    _Lexer.eof = false;
    _Lexer.line_number = 0;
    _Lexer.char_position = 0;

    // fill tokens and lexemes
    while ( index + length <= string.length ) {
        var small = string.substr( index, length ) ;
        var big = string.substr( index, length + 1 ) ;
        for ( var def in definitions ) {
            var smallmatch = small.match( definitions[def].regexp ) !== null;
            var bigmatch = big.match( definitions[def].regexp ) !== null;
            if ( smallmatch &&( ! bigmatch || small == big )  ) {
                // found a token
                index += length;
                length = 0;
                _Lexer.tokens.push(def);
                _Lexer.lexemes.push(small);
                break;
            }
        }
        length++;
    }

    _Lexer.next = function (count) {
        if (count == undefined) count = 1;
        while (count-- > 0) {
            arrayPosition++;
            if ( arrayPosition < _Lexer.tokens.length) {
                _Lexer.token = _Lexer.tokens[arrayPosition];
                _Lexer.lexeme = _Lexer.lexemes[arrayPosition];
                _Lexer.bof = false;
                _Lexer.char_position += _Lexer.lexeme.length;
                if (definitions[_Lexer.token].skip) {
                    if (_Lexer.token == 'newline') {
                        _Lexer.line_number++;
                        _Lexer.char_position = 0;
                    }
                    _Lexer.next();
                }
            } else {
                _Lexer.token = 'EOF';
                _Lexer.lexeme = null;
                _Lexer.eof = true;
            }
        }
    };

    _Lexer.prev = function (count) {
        if (count == undefined) count = 1;
        while (count-- > 0) {
            arrayPosition--;
            if ( arrayPosition-- > 0) {
                _Lexer.token = _Lexer.tokens[arrayPosition];
                _Lexer.lexeme = _Lexer.lexemes[arrayPosition];
                _Lexer.eof = false;
                _Lexer.char_position -= _Lexer.lexeme.length;
                if (definitions[_Lexer.token].skip) {
                    if (_Lexer.token == 'newline') {
                        _Lexer.line_number--;
                        _Lexer.char_position = 0;
                    }
                    _Lexer.prev();
                }
            } else {
                _Lexer.token = 'BOF';
                _Lexer.lexeme = null;
                _Lexer.bof = true;
                break;
            }
        }
    };
};

var grammar = {
    comment: { regexp: /^;.*$/, skip: true },
    number: { regexp: /^[1-9-]*[0-9][0-9./]*$/, skip: false },
    symbol: { regexp: /^[a-zA-Z_+*/=-]+[a-zA-Z0-9_+*/=|?<>-]*$/, skip: false },
    string: { regexp: /^\"[^\"]*\"$/, skip: false },
    quote: { regexp: /^'$/, skip: false },
    unquote: { regexp: /^~$/, skip: false },
    lparen: { regexp: /^\($/, skip: false },
    rparen: { regexp: /^\)$/, skip: false },
    whitespace: { regexp: /^[\t \n,]$/, skip: true }
};

module.exports = function lex(s) {
    var lexer = new Lexer(s, grammar);
    var stream = [];
    while (!lexer.eof) {
        lexer.next();
        stream.push({ type: lexer.token, value: lexer.lexeme });
    }
    return stream;
};

File: lex.py
from shlex import shlex
import re

class ParserError(Exception):
  def __init__(self, value, token = None):
    self.value = value
    self.token = token
  def __str__(self):
    if self.token:
      return "%s:%d: %s" % (self.token["infile"],
                            self.token["lineno"],
                            self.value)
    return self.value

classifiers = (
  ("lparen", re.compile(r"\(")),
  ("rparen", re.compile(r"\)")),
  ("string", re.compile(r"\".*")),
  ("quote", re.compile(r"'")),
  ("unquote", re.compile(r"~")),
  ("deref", re.compile(r"@")),
  ("number", re.compile(r"[1-9-]*[0-9][0-9./]*")),
  ("symbol", re.compile(r"[a-zA-Z_+*/=<>&|?!.-][a-zA-Z0-9_+*/=<>&|?!.-]*")),
)

def classify(token):
  for (t, r) in classifiers:
    if r.match(token):
      return t
  raise ParserError("Unclassifiable token \"%s\"" % token)

class Lexer(object):
  def __init__(self, stream, filename = None):
    if not filename and isinstance(stream, file):
      filename = stream.name
    l = shlex(stream, filename, posix = False)
    l.commenters = ";"
    l.wordchars = l.wordchars + "+-*/=<>&|?.!"
    l.whitespace = ", \t\r\n"
    l.escape = "\\"
    l.quotes = "\""
    self.lexer = l

  def __iter__(self):
    for token in self.lexer:
      yield { "type": classify(token),
              "value": token,
              "lineno": self.lexer.lineno,
              "infile": self.lexer.infile }
